File: wardenly-go/domain/scene/scene.go
// Package scene defines scene recognition types for visual automation.
package scene

import (
	"image"
	"image/color"
)

// Scene represents a recognizable game state defined by color points.
type Scene struct {
	// Name is the unique identifier for this scene
	Name string

	// Category groups related scenes (e.g., "city", "battle", "loading")
	Category string

	// Points are the color checkpoints used to identify this scene
	Points []Point

	// Actions are predefined actions available in this scene
	Actions map[string]Action
}

// Point represents a coordinate with an expected color for scene matching.
type Point struct {
	X     int
	Y     int
	Color color.RGBA
}

// Action represents a predefined action within a scene.
type Action struct {
	Type  ActionType
	Point ActionPoint
}

// ActionType represents the type of action.
type ActionType string

const (
	ActionTypeClick      ActionType = "click"
	ActionTypeWait       ActionType = "wait"
	ActionTypeDrag       ActionType = "drag"
	ActionTypeQuit       ActionType = "quit"
	ActionTypeIncr       ActionType = "incr"
	ActionTypeDecr       ActionType = "decr"
	ActionTypeCheckScene ActionType = "check_scene"
)

// ActionPoint represents coordinates for an action.
type ActionPoint struct {
	X float64
	Y float64
}

// Matcher provides scene matching functionality.
type Matcher struct {
	// Threshold is the maximum average color difference allowed for a match.
	// Lower values require more precise matches.
	Threshold float64
}

// NewMatcher creates a new scene matcher with the specified threshold.
func NewMatcher(threshold float64) *Matcher {
	if threshold <= 0 {
		threshold = 5.0 // Default threshold
	}
	return &Matcher{Threshold: threshold}
}

// Match checks if the given image matches this scene.
func (m *Matcher) Match(scene *Scene, img image.Image) bool {
	if len(scene.Points) == 0 || img == nil {
		return false
	}

	var totalDiff float64
	for _, point := range scene.Points {
		diff := colorDiff(img.At(point.X, point.Y), point.Color)
		totalDiff += diff
	}

	avgDiff := totalDiff / float64(len(scene.Points))
	return avgDiff <= m.Threshold
}

// MatchResult contains details about a scene match attempt.
type MatchResult struct {
	Scene      *Scene
	Matched    bool
	AvgDiff    float64
	PointDiffs []float64
}

// MatchWithDetails performs matching and returns detailed results.
func (m *Matcher) MatchWithDetails(scene *Scene, img image.Image) *MatchResult {
	result := &MatchResult{
		Scene:      scene,
		PointDiffs: make([]float64, len(scene.Points)),
	}

	if len(scene.Points) == 0 || img == nil {
		return result
	}

	var totalDiff float64
	for i, point := range scene.Points {
		diff := colorDiff(img.At(point.X, point.Y), point.Color)
		result.PointDiffs[i] = diff
		totalDiff += diff
	}

	result.AvgDiff = totalDiff / float64(len(scene.Points))
	result.Matched = result.AvgDiff <= m.Threshold

	return result
}

// colorDiff calculates the color difference between two colors.
// Returns a value between 0 (identical) and higher values for more difference.
func colorDiff(c1, c2 color.Color) float64 {
	r1, g1, b1, _ := c1.RGBA()
	r2, g2, b2, _ := c2.RGBA()

	// Convert from 16-bit to 8-bit
	r1, g1, b1 = r1>>8, g1>>8, b1>>8
	r2, g2, b2 = r2>>8, g2>>8, b2>>8

	// Calculate absolute differences
	dr := absDiff(r1, r2)
	dg := absDiff(g1, g2)
	db := absDiff(b1, b2)

	// Return average difference
	return float64(dr+dg+db) / 3.0
}

func absDiff(a, b uint32) uint32 {
	if a > b {
		return a - b
	}
	return b - a
}

File: wardenly-rs/src-tauri/src/domain/model/scene.rs
use image::DynamicImage;
use serde::{Deserialize, Serialize};

/// Scene represents a recognizable game state defined by color points.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Scene {
    /// Unique identifier for this scene
    pub name: String,

    /// Category groups related scenes (e.g., "city", "battle", "loading")
    #[serde(default)]
    pub category: String,

    /// Color checkpoints used to identify this scene
    pub points: Vec<ColorPoint>,

    /// Predefined actions available in this scene
    #[serde(default)]
    pub actions: std::collections::HashMap<String, SceneAction>,
}

/// ColorPoint represents a coordinate with an expected color for scene matching.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColorPoint {
    pub x: i32,
    pub y: i32,
    #[serde(flatten)]
    pub color: ColorValue,
}

/// Color value with optional tolerance
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ColorValue {
    pub r: u8,
    pub g: u8,
    pub b: u8,
    #[serde(default)]
    pub a: Option<u8>,
}

/// Predefined action within a scene
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SceneAction {
    #[serde(rename = "type")]
    pub action_type: String,
    pub point: Option<ActionPoint>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ActionPoint {
    pub x: f64,
    pub y: f64,
}

/// Scene matcher with configurable threshold
pub struct SceneMatcher {
    /// Maximum average color difference allowed for a match
    pub threshold: f64,
}

impl Default for SceneMatcher {
    fn default() -> Self {
        Self { threshold: 5.0 }
    }
}

impl SceneMatcher {
    pub fn new(threshold: f64) -> Self {
        Self {
            threshold: if threshold <= 0.0 { 5.0 } else { threshold },
        }
    }

    /// Check if the given image matches the scene
    pub fn matches(&self, scene: &Scene, image: &DynamicImage) -> bool {
        if scene.points.is_empty() {
            return false;
        }

        let rgb = image.to_rgb8();
        let (width, height) = (rgb.width(), rgb.height());

        let mut total_diff = 0.0;
        let mut valid_points = 0;

        for point in &scene.points {
            if point.x < 0 || point.y < 0 {
                continue;
            }

            let x = point.x as u32;
            let y = point.y as u32;

            if x >= width || y >= height {
                continue;
            }

            let pixel = rgb.get_pixel(x, y);
            let diff = color_diff(
                pixel[0],
                pixel[1],
                pixel[2],
                point.color.r,
                point.color.g,
                point.color.b,
            );
            total_diff += diff;
            valid_points += 1;
        }

        if valid_points == 0 {
            return false;
        }

        let avg_diff = total_diff / valid_points as f64;
        avg_diff <= self.threshold
    }

    /// Match with detailed results for debugging
    pub fn match_with_details(&self, scene: &Scene, image: &DynamicImage) -> MatchResult {
        let rgb = image.to_rgb8();
        let (width, height) = (rgb.width(), rgb.height());

        let mut point_diffs = Vec::with_capacity(scene.points.len());
        let mut total_diff = 0.0;
        let mut valid_points = 0;

        for point in &scene.points {
            if point.x < 0 || point.y < 0 {
                point_diffs.push(f64::MAX);
                continue;
            }

            let x = point.x as u32;
            let y = point.y as u32;

            if x >= width || y >= height {
                point_diffs.push(f64::MAX);
                continue;
            }

            let pixel = rgb.get_pixel(x, y);
            let diff = color_diff(
                pixel[0],
                pixel[1],
                pixel[2],
                point.color.r,
                point.color.g,
                point.color.b,
            );
            point_diffs.push(diff);
            total_diff += diff;
            valid_points += 1;
        }

        let avg_diff = if valid_points > 0 {
            total_diff / valid_points as f64
        } else {
            f64::MAX
        };

        MatchResult {
            scene_name: scene.name.clone(),
            matched: avg_diff <= self.threshold,
            avg_diff,
            point_diffs,
        }
    }
}

/// Result of a scene match attempt
#[derive(Debug, Clone)]
pub struct MatchResult {
    pub scene_name: String,
    pub matched: bool,
    pub avg_diff: f64,
    pub point_diffs: Vec<f64>,
}

/// Calculate average color difference between two colors
fn color_diff(r1: u8, g1: u8, b1: u8, r2: u8, g2: u8, b2: u8) -> f64 {
    let dr = (r1 as i32 - r2 as i32).abs() as f64;
    let dg = (g1 as i32 - g2 as i32).abs() as f64;
    let db = (b1 as i32 - b2 as i32).abs() as f64;
    (dr + dg + db) / 3.0
}

impl Scene {
    /// Convenience method to check if image matches this scene using default threshold
    pub fn matches(&self, image: &DynamicImage) -> bool {
        SceneMatcher::default().matches(self, image)
    }
}

